app/views/courses/TeacherClassView.coffee

```diff
@@ -390,35 +386,92 @@ module.exports = class TeacherClassView extends RootView
   onClickBulkAssign: ->
     courseID = @$('.bulk-course-select').val()
     selectedIDs = @getSelectedStudentIDs()
-    members = selectedIDs.filter (userID) =>
-      user = @students.get(userID)
-      user.isEnrolled()
-    assigningToUnenrolled = _.any selectedIDs, (userID) =>
-      not @students.get(userID).isEnrolled()
     assigningToNobody = selectedIDs.length is 0
-    @state.set errors: { assigningToNobody, assigningToUnenrolled }
+    @state.set errors: { assigningToNobody }
     return if assigningToNobody
-    @assignCourse courseID, members
+    @assignCourse courseID, selectedIDs
     window.tracker?.trackEvent 'Teachers Class Students Assign Selected', category: 'Teachers', classroomID: @classroom.id, courseID: courseID, ['Mixpanel']
 
-  # TODO: Move this to the model. Use promises/callbacks?
   assignCourse: (courseID, members) ->
-    courseInstance = @courseInstances.findWhere({ courseID, classroomID: @classroom.id })
-    if courseInstance
-      courseInstance.addMembers members
-    else
-      courseInstance = new CourseInstance {
-        courseID,
-        classroomID: @classroom.id
-        ownerID: @classroom.get('ownerID')
-        aceConfig: {}
-      }
-      @courseInstances.add(courseInstance)
-      courseInstance.save {}, {
-        success: ->
-          courseInstance.addMembers members
-      }
-    null
+    courseInstance = null
+    numberEnrolled = 0
+    remainingSpots = 0
+    
+    return Promise.resolve()
+    .then =>
+      courseInstance = @courseInstances.findWhere({ courseID, classroomID: @classroom.id })
+      if not courseInstance
+        courseInstance = new CourseInstance {
+          courseID,
+          classroomID: @classroom.id
+          ownerID: @classroom.get('ownerID')
+          aceConfig: {}
+        }
+        courseInstance.notyErrors = false # handling manually
+        @courseInstances.add(courseInstance)
+        return courseInstance.save()
+        
+    .then =>
```

Comment:
If there's no course instance, it returns a `jqxhr`. Otherwise it returns `undefined`. It never returns anything "promise-wrapped", it's either something with a `then` function attached or not. If this function returns the `jqxhr`, the Promise chain recognizes it as a Promise because it has its own `then` function, and waits for it to finish then continues. Otherwise, the Promise chain continues immediately (or very soon).